### ISSUE TITLE
Make retry customizable

### ISSUE THREAD

**Issue by lucaswerkmeister** [lucaswerkmeister/m3api#24]
Since I’m fairly sure I want to put OAuth support (#9) into an extension package and not into m3api core, but OAuth 2.0 access tokens are relatively short-lived (4 h) and need to be refreshed from time to time, m3api needs to let extension packages hook into the retry mechanism.

**Comment by lucaswerkmeister** [comment]
I think we can generalize the current retry code in `internalRequest()`, based on the error code. The options could contain an optional mapping from error code to retry function (or better yet, array of retry functions, so multiple packages can integrate with each other more easily). The function would return an optional promise; if the promise resolves to `true`, then retry the request. (Modulo some modifications related to the “only retry if we have enough time left” logic, which I haven’t fully figured out yet.)

But there’s one problem with this. Currently, `internalRequest()` gets the request options in resolved form, and calls itself to retry – e.g. changing the session’s `method` option won’t affect the retry of ongoing requests, even if those requests didn’t specify an explicit `method`. But the OAuth package will need to update the `authorization` header/option of the requests for retry.

We could somehow shoehorn that header into the retrier promise response, but that feels inelegant to me. I think it might be better to make `internalRequest()` call `request()`, rather than itself, for retry; then the `authorization` option would automatically be re-resolved in the retry `request()` when it resolves all the options again. In 93690038efafce05f254a9ca93249262b5d9e9af, I wrote that “I don’t want to make internalRequest() call request() on retry”, but without giving a reason; I think it might mainly have been that this would require turning the `retryUntil` timestamp back into a `maxRetriesSeconds` duration option. But I now think it might be a better tradeoff to make `internalRequest()` call `request()` on retry after all.

If `internalRequest()` calls `request()` on retry, I think that actually makes the behavior of changing params and options more consistent. Right now, if you start a request and then immediately change the session’s default params and options, for the most part those changes will still apply to that request after it’s (potentially) been combined, if I’m not mistaken; so it seems consistent to also make those changes apply to retries. (Basically, you should be cautious when you’re changing the default params and options anyways.)

**Comment by lucaswerkmeister** [comment]
> I think it might mainly have been that this would require turning the `retryUntil` timestamp back into a `maxRetriesSeconds` duration option.

I think we should avoid this by adding a request option corresponding to `retryUntil`, which is not part of the public interface but takes precedence over `maxRetriesSeconds` if present. (Not sure if `retryUntil` is the best name for it.) If we convert the limit back and forth between absolute and relative time, it’ll drift a bit.

> The options could contain an optional mapping from error code to retry function (or better yet, array of retry functions, so multiple packages can integrate with each other more easily).

I’m second-guessing that “better yet” now. On the one hand, it does give packages more control about how their retry functions interact: in an array, they can insert themselves in any position or even reorder or remove other packages’ functions, whereas with a single functions their only options are to chain themselves before the accumulated function so far, after it, or to overwrite it without chaining.

On the other hand, supporting an array of retry functions significantly complicates the logic to combine options. Instead of just splatting the options together with the spread operator, letting more specific options overwrite more general ones, libraries would probably expect that the arrays of retry functions are *merged*, with the more specific options first. We’d probably need a new function implementing this merging logic in m3api, and require extension packages to call it instead of just copy+pasting the simple code snippet `{ ...DEFAULT_OPTIONS, ...session.defaultOptions, ...options }`. And while that’s doable, I highly doubt it’s worth it: I don’t actually have a use case in mind where any packages would want to manipulate each others’ retry retry functions. (`maxlag`, `readonly`, `badtoken` retry would stay in `m3api`, and I don’t see a reason why extension packages would want to change them; `mwoauth-invalid-authorization` would be exclusive to `m3api-oauth2`.)

**Comment by lucaswerkmeister** [comment]
> (Modulo some modifications related to the “only retry if we have enough time left” logic, which I haven’t fully figured out yet.)

I think there’s at least two approaches here. The retry function could return a number, indicating to sleep for this many seconds unless that would exceed the retry limit; but also, the retry function will definitely need to receive the `session` and `options` as arguments (so that `m3api-oauth2` can make a request to refresh the access token), and if those options include the `retryUntil` option (see previous comment; name not guaranteed), the retry function can decide for itself whether there’s enough time for the sleep or not. (The option isn’t planned to be part of the stable interface, but also, so far only retry functions inside m3api itself would need it, for `maxlag` and `readonly` errors.)

Returning a number is a bit opaque (does it represent seconds, milliseconds, microseconds, something else?), so maybe just relying on the option is better.

**Comment by lucaswerkmeister** [comment]
> On the other hand, supporting an array of retry functions significantly complicates the logic to combine options.

Bleh, I forgot that the logic is actually complicated regardless of whether we support arrays of retry functions or single retry functions, because in either case we already have one level of nesting because the retry functions are per error code.

But I also realize now: it’s not the worst thing in the world if the merge logic gets a bit more complicated. After all, we don’t want extension packages to pass merged options through to m3api: they’re only supposed to merge options for their own extension package options, and when they pass through options to m3api, use the original options for that. So it’s perfectly fine to process the options like this in m3api:

```js
const {
	method,
	tokenType,
	tokenName,
	maxRetries, // only for warning
	maxRetriesSeconds,
	retryAfterMaxlagSeconds,
	retryAfterReadonlySeconds,
	warn,
	dropTruncatedResultWarning,
	authorization,
} = {
	...DEFAULT_OPTIONS,
	...this.defaultOptions,
	...options,
};
const retryPerErrorCode = {
	...DEFAULT_OPTIONS.retryPerErrorCode,
	...this.defaultOptions.retryPerErrorCode,
	...options.retryPerErrorCode,
};
```

And extension packages continue to process their own options like this:
```js
const {
	'm3api-query/mergeValues': mergeValues,
	'm3api-query/comparePages': comparePages,
} = {
	...DEFAULT_OPTIONS,
	...session.defaultOptions,
	...options,
};
```
And they don’t need to care that the `retrytPerErrorCode` option is more complicated than that.

I’ll still go for a single retry function, though – that’s still simpler than having to concat the arrays, and the argument that we can defer that complexity until we need it still stands.

**Comment by lucaswerkmeister** [comment]
Perhaps the “retry handlers” could actually be “error handlers”? Something like:

- The error handler can return a Promise, which m3api will await.
- If it returns `true` or resolves to `true`, m3api will retry the request.
  - Note: For a moment, I thought we could do the “is there enough time for a retry?” logic in the general error handler code: after the promise resolved to true, only do the retry if we’re not over the time limit. But that’s no good: if we’re going to sleep for a certain time before retrying, we don’t want to sleep if we won’t have enough time for the retry anyway. So the specific error handler needs to decide if, in the promise it returns, it should sleep and return `true` or directly return `null` without sleeping.
- If it returns `false` or resolves to `false`, m3api will consider the error handled and remove it from the list of errors, but not retry the request.
- If it returns `null` or resolves to `null`, m3api will not consider the error handled and keep it in the list of errors (and not retry the request).
- If it throws or rejects, m3api will not consider the error handled, not retry, and… do what with the rejection? Throw that instead of the API errors?
- After calling all error handlers (assuming none of them triggered a retry), if there are any errors left, throw them; otherwise return the response. (The errors would still be contained in the response in that case.)

It’s a bit awkward if one error handler returns `null` (error not handled) and then a subsequent error handler returns `true` (retry), meaning that the first error handler’s unhandled error still gets discarded, but I don’t think there’s a better way to handle this situation generically. (Also, I think it’s worth keeping in mind that while MediaWiki *can* return multiple errors in formatversion=2, many cases will still just produce a single error.)

You could also argue that returning `true`/`false`/`null` for this isn’t very readable, and perhaps an object like `{ handled: true, retry: true }` would be clearer.

Also, is there actually a use case for a handled error without retry (`false` above)?

**Comment by lucaswerkmeister** [comment]
> Note: For a moment, I thought we could do the “is there enough time for a retry?” logic in the general error handler code: after the promise resolved to true, only do the retry if we’re not over the time limit. But that’s no good: if we’re going to sleep for a certain time before retrying, we don’t want to sleep if we won’t have enough time for the retry anyway. So the specific error handler needs to decide if, in the promise it returns, it should sleep and return `true` or directly return `null` without sleeping.

We can actually take this one step further and say that the error handler is *fully* responsible for respecting the `retryUntil` option; i.e., a synchronous error handler, or one that makes another request instead of sleeping for a set period of time (renew OAuth access token), would also be expected to check whether `retryUntil` has passed or not, and m3api itself just does the retry whenever the error handler tells it to, regardless of the `retryUntil` option.

And then this morning I realized another thing. We’re already passing the session and request options into the handler, so that `m3api-oauth2` can make the request to renew the access token. If we also pass in the request params, then the handler has all the information it needs to just make the retry request itself. Under this idea, m3api would just return whatever promise the handler returned, and the handler would return `session.request( params, options )` if it determines the request should be retried.

However, we still need a provision for *not* retrying a request in this case – and I think this provision even needs to be asynchronous: if `m3api-oauth2` renews the access token, but then the `retryUntil` time has passed when that request finishes, it should be able to indicate to m3api that the original request should still be considered an error because there’s not enough time to retry the request now (even if it looked like there might be enough time before we started the request to renew the access token).

Assuming that there *isn’t* a use case for a handled error without retry (`false` or `{ handled: true, retry: false }` in the scheme outlined in the previous comment), I think we can fall back to “truthy/falsy” here: the handler returns a value, or a promise resolving to a value; if the value is truthy (or more specifically an object?), that’s the result of the request; if the value is falsy, the error wasn’t handled and we should throw the API errors at the end.

**Comment by lucaswerkmeister** [comment]
> If we also pass in the request params, then the handler has all the information it needs to just make the retry request itself.

This is true, but I forgot that m3api *also* retries in the presence of a `Retry-After` response error, regardless of any errors. So I need to fit this in somehow as well.
